### DISCUSSION TITLE
Set the maximum keyframes to avoid memory problems

### DISCUSSION THREAD

**Discussion by abylikhsanov** [stella-cv/stella_vslam#214] [Q&A] [answered]
Hi,

Is there a way to set a maximum number of keyframes or landmarks to set the maximum RAM size that is usable for SLAM? For example deleting old keyframes if certain threshold has exceeded and etc.

**Comment by ymd-stella** [comment, marked as answer]
No, it doesn't.
In addition, in the main branch, setting an upper limit on the number of keyframes does not reduce memory usage.

If high memory usage is your problem, I recommend you to use use-shared-ptr branch.
In this branch, unused keyframes are released.
The keyframe limit is not implemented yet, but if you create an issue for a feature request, someone may implement it when they feel like it. Your pull requests are also welcome.

(Note: I just merged use-shared-ptr into main.)

**Reply by abylikhsanov** [reply]
Thanks. I have a question though. If I have a certain origin during the initial mapping phase and then just keep extending the map by running the localizaition with mapping argument, will the origin be possibly shifted overtime because the newly merged branch use-shared-ptr will at some point might release the keyframes at the origin?

**Reply by ymd-stella** [reply]
Only keyframes removed from map_database by local_map_cleaner will be removed from memory. See https://github.com/OpenVSLAM-Community/openvslam/blob/61db7959cf6deb093f543bbd3d313868edd869d4/src/openvslam/module/local_map_cleaner.cc#L72-L105

**Reply by abylikhsanov** [reply]
Thanks. Seems that limiting the amount of new keyframes is quite easy to add and the origin is already "frozen"

**Reply by abylikhsanov** [reply]
Hi, so I have started to investigate this and I think it would be quite benefitial for many as long term SLAM when robot should operate without any interruption (mostly commercial grade industrial robots, like AMRs).

So I would like to outline my brief implementation idea below and would like to hear your opinion on this:

1. After the initial map, store the id of the last keyframe.
2. Next time, when loading a map and extending it, take a note of the last keyframe and keep track of the available memory left and if some threshold has reached (I can check this in tracking module at KF decision), I do not add any keyframes but rather delete half keyframes that are added from last keyframe (from 1) until current, change the id of the current potential keyframe to follow the correct sequence and continue the GO and mapping procedures as usual

The reason I want to delete the first half is quite naive - because it is better to delete the older added keyframes to keep the memory under control.

**Reply by ymd-stella** [reply]
> I have started to investigate this and I think it would be quite benefitial for many as long term SLAM when robot should operate without any interruption (mostly commercial grade industrial robots, like AMRs).

How much RAM do you expect to have?

> I would like to outline my brief implementation idea below and would like to hear your opinion on this

Define the problem properly first, not the implementation idea.

First, is the target space large or small?
If it is a large space, deleting old keyframes will cause accumulation of errors, which is closer to Visual Odometry than Visual SLAM. In this case, if RAM is limited, you should consider other means (like Visual Odometry) rather than OpenVSLAM.
For a small space, the same thing can be achieved by simply adjusting the keyframe insertion/removal algorithms or parameters.

By process of elimination, I would consider using OpenVSLAM for small spaces. This is the problem of how to make insertion/deletion decisions for an upper bounded RAM. What you have proposed is one of the solutions. However, not only keyframes should be limited, but also landmarks and feature points. In particular, I think the strategy of inserting and removing keyframes and landmarks should be considered as a research first. If you haven't done a paper survey yet, why not do it?

If you find (or create) a sufficiently sophisticated strategy and implement it, I will merge it so that it can be selected. If you do, please submit a pull request.

**Reply by ymd-stella** [reply]
In addition, the computational resources of the robot will increase in the future. For example, Jetson AGX xavier has increased its memory from 16GB to 64GB in a few years. It is easier to solve the problems with hardware. This is why I am not very interested in this feature.

**Reply by abylikhsanov** [reply]
> How much RAM do you expect to have?

32 GB

> Define the problem properly first, not the implementation idea.

Problem is that no matter how big is your RAM, in real world robots, the function of continiously updating the map will eventually make any robot to run out of memory (it might take months but that is still quite a short period of time). The good thing is that these robots usually operate in one environment so the map is sound and confined (yes you will have dynamic objects here and there but still many features on the walls are quite static that does not require constant updates and those dynamic objects are also finite meaning that new objects will be present but old objects will also be gone so we can free up the memory).

> First, is the target space large or small?
> If it is a large space, deleting old keyframes will cause accumulation of errors, which is closer to Visual Odometry than Visual SLAM

Large or small is quite a subjective and relative description. What I meant is that there is a way to calculate how much memory you have consumed to map a certain area so in theory, it is possible to predict what would it take for a robot to keep mapping the area with at least an additional full loop of the same map (for example, if it took you around 4GB to map the area, in theory covering the same map at least once would probably take 8GB in worst case,  I might be wrong here as several loop closures do take quite a significant amount of memory). 
Also, I do not want to delete the keyframes that were present during the first initial map, only the subsequently added keyframes/landmarks after that so it should not case accumulation of errors.

> In addition, the computational resources of the robot will increase in the future. For example, Jetson AGX xavier has increased its memory from 16GB to 64GB in a few years. It is easier to solve the problems with hardware. This is why I am not very interested in this feature.

That would increase the cost quite significantly as memory is one of the most expensive hardware component and again, even if you would insert 128GB, still you have a problem of never ending growth which is just scary thing for a commercial robot.
The reason I am interested in this, is because I believe this is the future of all commercial robots, VSLAM outperforms other SLAM methods when the environment is complex and dynamic and with this constraint memory and never ending learning functionality, it will be just a killer product.

I will investigate this problem and will probably start with the simple idea to understand the impact

.

**Reply by ymd-stella** [reply]
> (for example, if it took you around 4GB to map the area, in theory covering the same map at least once would probably take 8GB in worst case, I might be wrong here as several loop closures do take quite a significant amount of memory).

This assumption is incorrect. Unnecessary keyframes will not be inserted, and redundant keyframes will be deleted. Therefore, unless the view of the room changes significantly, the actual memory usage is approximately 4GB. Or is there another assumption that you have not yet made clear?

**Reply by ymd-stella** [reply]
> Large or small is quite a subjective and relative description.

No, it is relative, but not subjective. It asks if the map will fit on the limited memory. I'll ask again. Are you assuming that the target is the space where the map will fit in memory?

**Reply by ymd-stella** [reply]
Redefine the problem. I wish you would have sorted this much out before posting.

- Use a machine with a 32GB RAM limit.
- We will do Visual SLAM in a space the size of which will fit in memory if mapped normally.
- Objects in the space are dynamic, so we need to keep mapping.

Is this enough? Are there any other assumptions you're forgetting?

**Reply by abylikhsanov** [reply]
> This assumption is incorrect. Unnecessary keyframes will not be inserted, and redundant keyframes will be deleted. Therefore, unless the view of the room changes significantly, the actual memory usage is approximately 4GB. Or is there another assumption that you have not yet made clear?


The actual environment is the same but objects which are up to 4-5 meters in height (for example in warehouse) always change so ideally, the openvslam should note that some objects (like pallets/boxes) have been observed less than observable and with low enough ratio the local map cleaner should be able to delete the keyframes and landmarks. In reality, I did not observe this. Based on my rosbags, the amount of keyframes were only increasing and the amount of keyframes and landmarks that are deleted were very insignificant. It looked like if I would supply long enough rosbag it will eventually add more than deleted.

> Redefine the problem. I wish you would have sorted this much out before posting.
> 
> Use a machine with a 32GB RAM limit.
> We will do Visual SLAM in a space the size of which will fit in memory if mapped normally.
> Objects in the space are dynamic, so we need to keep mapping.
> Is this enough? Are there any other assumptions you're forgetting?

- Use a machine with a 32GB RAM limit.
- We will initially do Visual SLAM in a space with the size of which will fit in a memory if mapped normally. This map will eventually be saved
- We will then load the saved map with argument --mapping which will enable us to continue the mapping
- Objects in the space are dynamic, so we need to always add/delete keyframes,landmarks and possibly keypoints
- There should be a physical limit for adding new keyframes and landmarks for a safety reasons when the amount of used memory is close to 32 GB. In this case, we should either delete the oldest newly added keyframes since localization mode with mapping argument or disable addition for a period of time
- We should avoid error accumulation therefore we should disable erase of any objects that were made during the initial visual SLAM procedure.

**Reply by ymd-stella** [reply]
It is not clear why simply disabling mapping is not enough.

It would be better to map without dynamic objects, or to build up the map until the dynamic objects are no longer affected, then stop mapping and only track.

**Reply by abylikhsanov** [reply]
Yes that would be ideal of course but in reality, it is not possible. If you take a large warehouse for logistic company, it is simply not feasible to stop the ongoing business activity (forklifts passing by, moving pallets, boxes) and also take away all pallets and boxes from the warehouse shelves which can be more than thousands. 

So in this case, you can actually map as it is and then just disable mapping module but then you would need some quite good amount of features that are static, for example the ceiling but that limits the robustness when shelves are tall enough and only small FoV is the ceiling.

**Reply by ymd-stella** [reply]
Is the survey sufficient? There are probably simpler methods for localization using only the ceiling.
